### DISCUSSION TITLE
Moshi kapt deprecation warning always shown?

### DISCUSSION THREAD

**Discussion by G00fY2** [square/moshi#1752] [Q&A] [answered]
Does the deprecation warning about kapt is always shown, regardless of Moshi already being used with ksp?

https://github.com/square/moshi/blob/491e66bd6475860ec0cba35f76dd14ec88fbeeca/moshi-kotlin-codegen/src/main/java/com/squareup/moshi/kotlin/codegen/apt/JsonClassCodegenProcessor.kt#L73

**Comment by ZacSweers** [comment, marked as answer]
If it's being shown, it means Moshi is being loaded into kapt/apt and you should remove it from that configuration.

**Reply by G00fY2** [reply]
Because we 100% removed all KAPT usage in our project (Hilt, Room, Moshi now all supporting KSP) but we still see this warning.

But only for the `:app:hiltJavaCompile` task. Is Hilt (1.1.0-rc01) maybe internally using Moshi or something?

I also checked our dependency graph. Only `com.squareup.retrofit2:converter-moshi` has a runtime dependency on Moshi, which should not affect KAPT/KSP.

Any other ideas how to find out what causes this warning in our project?

**Reply by ZacSweers** [reply]
If you're using gradle, you can run the `dependencies` task on that project and inspect what is bringing the Moshi code gen dependency to the kapt classpath. If you're not using kapt at all (I.e. the kapt plugin isn't even applied), I'm not sure how this could happen and would want to see a public repro project

**Reply by G00fY2** [reply]
Hi @ZacSweers, thanks. Here is a repro project where I just added moshi and dagger/hilt. In the logs you can see the error message:
https://github.com/G00fY2/HiltMoshiKaptWarning/actions/runs/6683702355/job/18160177116#step:5:44

Should I maybe create an issue in the dagger repo?

**Reply by ZacSweers** [reply]
This is a hilt bug, it's loading all annotation processors for some reason in the `:app:hiltJavaCompileDebug` task. Looks like it's capturing all KSP and kapt dependencies and passing them on to the JavaCompile task it creates

<img width="1505" alt="image" src="https://github.com/square/moshi/assets/1361086/0f4fad88-87a5-44bb-a270-e7bb6b19644a">

**Reply by G00fY2** [reply]
I created https://github.com/google/dagger/issues/4116. Let's see what the dagger team says.

**Reply by mkubiczek** [reply]
How to suppress this warning when `-Werror` compiler flag is used?

**Reply by G00fY2** [reply]
From my perspective the easiest solution would be changing moshis warning messages, as suggested by the hilt team.

Not sure how the Moshi maintainers feel about it? Maybe it's worth creating an issue and/or a PR to address this?

> Moshi could change the location where the message is reported such that only if it really needs to do processing then (i.e. in a round) it reports the warning and not when it is loaded. This is up to the Moshi authors to decide if it is possible and still maintains their desired migration to KSP goals.

**Reply by ZacSweers** [reply]
Hilt is loading KSP dependencies into an APT classpath. Tbh I don't see why annotation processor maintainers need to do something to work around hilt's behavior

**Comment by jdelga** [comment]
It has been more than one year with this _annoying noisy warning_ although kapt is not used in the project.

Is there any update on a 2.0 version?

**Reply by ZacSweers** [reply]
And for over a year it's been a hilt bug: https://github.com/square/moshi/discussions/1752#discussioncomment-7417741. Coming here to whine about something you could ultimately resolve for yourself a number of ways is unconstructive.

**Reply by jdelga** [reply]
The message actually reads `Kapt support in Moshi Kotlin Code Gen is deprecated and will be removed in 2.0`

I am just asking if there is such version 2.0 in the near future.

Thanks.

**Reply by TWiStErRob** [reply]
Anyone who watches: KAPT code has been removed in https://github.com/square/moshi/pull/1848 in 2024 June, and that's released in [2.0.0-alpha.1](https://github.com/square/moshi/releases/tag/2.0.0-alpha.1) 2026 January.

**Reply by muzso** [reply]
> Coming here to whine about something you could ultimately resolve for yourself a number of ways is unconstructive.

Can you give us a couple of hints on how we can resolve this for ourselves? I'm new both to Moshi and Hilt. I've tried to google the solution, which led me here.

So far it seems to me that one the following has to happen:
- wait for a patch from Hilt
- wait for Moshi 2.* to be released
- fork either Moshi or Hilt and fix the issue ourselves (submitting a PR seems illogical since maintainers of both projects consider the issue to be with the other project)

The first two might happen at some point in the future (they don't resolve the warning for an indefinite amount of time), the third is not really an option (the effort of maintaining a locally patched version of either of these two projects would outweigh the benefits of getting rid of the given warning).

**Reply by muzso** [reply]
I googled a bit more (read a number of GH issues and their comments, a number of StackOverflow questions, etc.), but still no solution/workaround in sight.

Eventually I settled with asking ChatGPT (since I'm not an Android programmer, it was unlikely that I'd just find a solution by myself within reasonable time) and with a bit of back and forth we got to a solution. From what I can tell, it's exactly the kind of solution that @ZacSweers was referring to. I.e. removing Moshi from Hilt's (i.e. the `hiltJavaCompileDebug` task's) classpath.

I've added the following to the end of my `app/build.gradle.kts` and the warning is gone:
```
tasks.withType<JavaCompile>().configureEach {
  if (name.startsWith("hiltJavaCompile")) {
    doFirst {
      options.annotationProcessorPath =
        options.annotationProcessorPath?.filter {
          !it.name.startsWith("moshi-kotlin-codegen-")
        }
    }
  }
}
```
(Note: you'll need an `import org.gradle.api.tasks.compile.JavaCompile` too, preferably at the start of the file)

As stated before: I'm not an Android developer. I had no idea why the "doFirst { ... }" is required, but I've tried it without `doFirst` and `options.annotationProcessorPath` seems to be null and the Moshi warning remains.

Asked ChatGPT about it and got the following reason: "It's required here because Hilt doesn't populate hiltJavaCompileDebug's annotationProcessorPath until later in the build lifecycle."

So that's the full explanation to newbies like me.
